runtime/src/parachains.rs

```diff
@@ -248,6 +251,39 @@ impl<T: Trait> Module<T> {
 			.collect())
 	}
 
+	fn check_egress_queue_roots(head: &AttestedCandidate, active_parachains: &[ParaId]) -> Result {
+		let mut last_egress_id = None;
+		let mut iter = active_parachains.iter();
+		for (egress_para_id, root) in &head.candidate.egress_queue_roots {
+			// egress routes should be ascending order by parachain ID without duplicate.
+			ensure!(
+				last_egress_id.as_ref().map_or(true, |x| x < &egress_para_id),
+				"Egress routes out of order by ID"
+			);
+
+			// a parachain can't route to self
+			ensure!(
+				*egress_para_id != head.candidate.parachain_index,
+				"Parachain routing to self"
+			);
+
+			// no empty trie roots
+			ensure!(
+				*root != [EMPTY_TRIE; 32].into(),
```

Comment:
`[EMPTY_TRIE; 32]` is just `[0; 32]`?

By the root of the empty trie I mean that this would be `hash(empty_node)` which is dependent on the hash algorithm and node codec. I think the `substrate-trie` crate exposes this value as a constant. If it doesn't, it should expose a way to get the empty trie root and we should use it here (although memoizing so we only do the actual hashing once)